pkg/ipamclaimswebhook/podmutator.go

```diff
@@ -204,3 +170,117 @@ func podWithUpdatedSelectionElements(pod *corev1.Pod, networks []v1.NetworkSelec
+	log := logf.FromContext(ctx)
+	hasChangedNetworkSelectionElements := false
+	for i, networkSelectionElement := range networkSelectionElements {
+		nadName := types.NamespacedName{
+			Namespace: networkSelectionElement.Namespace,
+			Name:      networkSelectionElement.Name,
+		}.String()
+		log.Info(
+			"iterating network selection elements",
+			"NAD", nadName,
+		)
+		nadKey := types.NamespacedName{
+			Namespace: networkSelectionElement.Namespace,
+			Name:      networkSelectionElement.Name,
+		}
+
+		nad := v1.NetworkAttachmentDefinition{}
+		if err := cli.Get(context.Background(), nadKey, &nad); err != nil {
+			if k8serrors.IsNotFound(err) {
+				return false, "NAD not found, will hang on scheduler", nil
+			}
+			return false, "", err
+		}
+
+		pluginConfig, err := config.NewConfig(nad.Spec.Config)
+		if err != nil {
+			return false, "", err
+		}
+
+		if !pluginConfig.AllowPersistentIPs {
+			continue
+		}
+
+		log.Info(
+			"will request persistent IPs",
+			"NAD", nadName,
+			"network", pluginConfig.Name,
+		)
+
+		vmKey := types.NamespacedName{Namespace: pod.Namespace, Name: vmName}
+		vmi := &virtv1.VirtualMachineInstance{}
+		if err := cli.Get(context.Background(), vmKey, vmi); err != nil {
+			return false, "", err
+		}
+
+		vmiNets := vmiSecondaryNetworks(vmi)
+		networkName, foundNetworkName := vmiNets[nadKey.String()]
+		if !foundNetworkName {
+			log.Info(
+				"network name not found",
+				"NAD", nadName,
+				"network", networkName,
+			)
+			continue
+		}
+
+		networkSelectionElements[i].IPAMClaimReference = claims.GenerateName(vmName, networkName)
+		log.Info(
+			"requesting claim",
+			"NAD", nadName,
+			"network", pluginConfig.Name,
+			"claim", networkSelectionElement.IPAMClaimReference,
+		)
+		hasChangedNetworkSelectionElements = true
+		continue
+	}
+	return hasChangedNetworkSelectionElements, "", nil
+}
+
+func findNewPrimaryNetworkIPAMClaimName(ctx context.Context,
+	cli client.Client, pod *corev1.Pod, vmName string) (string, error) {
+	log := logf.FromContext(ctx)
+	if pod.Annotations[config.OVNPrimaryNetworkIPAMClaimAnnotation] != "" {
+		return "", nil
+	}
+	primaryNetworkNAD, err := udn.FindPrimaryNetwork(ctx, cli, pod.Namespace)
+	if err != nil {
+		return "", err
+	}
+	if primaryNetworkNAD == nil {
+		return "", nil
+	}
+	pluginConfig, err := config.NewConfig(primaryNetworkNAD.Spec.Config)
+	if err != nil {
+		return "", err
+	}
+
+	if !pluginConfig.AllowPersistentIPs {
+		return "", nil
+	}
+
+	log.Info(
+		"will request primary network persistent IPs",
+		"NAD", client.ObjectKeyFromObject(primaryNetworkNAD),
+		"network", pluginConfig.Name,
+	)
+	vmKey := types.NamespacedName{Namespace: pod.Namespace, Name: vmName}
+	vmi := &virtv1.VirtualMachineInstance{}
+	if err := cli.Get(context.Background(), vmKey, vmi); err != nil {
+		return "", err
+	}
+
+	networkName := vmiPodNetworkName(vmi)
+	if networkName == "" {
+		log.Info("vmi has no pod network primary UDN ipam claim will not be created")
```

Comment:
please print the namespaced name of the VMI that doesn't have a pod network name.